client/src/main/java/org/eclipse/hono/client/impl/AmqpAdapterClientCommandResponseSender.java

```diff
@@ -0,0 +1,131 @@
+
+        Objects.requireNonNull(con);
+        Objects.requireNonNull(tenantId);
+
+        return con.createSender(null, ProtonQoS.AT_LEAST_ONCE, remoteCloseHook)
+                .map(sender -> new AmqpAdapterClientCommandResponseSender(con, sender, tenantId));
+    }
+
+    @Override
+    public Future<ProtonDelivery> sendCommandResponse(final String deviceId, final String targetAddress,
+            final String correlationId, final int status, final byte[] payload, final String contentType,
+            final Map<String, ?> properties) {
+        return sendCommandResponse(deviceId, targetAddress, correlationId, status, payload, contentType,
+                properties, null);
+    }
+
+    @Override
+    public Future<ProtonDelivery> sendCommandResponse(final String deviceId, final String targetAddress,
+            final String correlationId, final int status, final byte[] payload,
+            final String contentType, final Map<String, ?> properties, final SpanContext context) {
+
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(targetAddress);
+        Objects.requireNonNull(correlationId);
+
+        final Message message = ProtonHelper.message();
+        message.setAddress(targetAddress);
+        message.setCorrelationId(correlationId);
+
+        MessageHelper.setCreationTime(message);
+        MessageHelper.addProperty(message, MessageHelper.APP_PROPERTY_TENANT_ID, tenantId);
+        MessageHelper.addProperty(message, MessageHelper.APP_PROPERTY_DEVICE_ID, deviceId);
+        MessageHelper.addProperty(message, MessageHelper.APP_PROPERTY_STATUS, status);
+        MessageHelper.setPayload(message, contentType, payload);
+
+        setApplicationProperties(message, properties);
+
+        return sendAndWaitForOutcome(message, context);
+    }
+
+    /**
+     * This method must not be used outside a protocol adapter.
+     * 
+     * @param commandResponse not used.
+     * @param context not used.
+     * @return nothing.
+     * @throws UnsupportedOperationException always.
+     */
+    @Override
+    public Future<ProtonDelivery> sendCommandResponse(final CommandResponse commandResponse,
+            final SpanContext context) {
+        throw new UnsupportedOperationException("This method must not be used outside a protocol adapter");
+    }
+
+    /**
+     * Shadowing the static factory method that must not be used outside a protocol adapter.
+     *
+     * @param con not used.
+     * @param tenantId not used.
+     * @param replyId not used.
+     * @param closeHook not used.
+     * @return nothing.
+     * @throws UnsupportedOperationException always.
+     */
+    public static Future<CommandResponseSender> create(
+            final HonoConnection con,
+            final String tenantId,
+            final String replyId,
+            final Handler<String> closeHook) {
+        throw new UnsupportedOperationException("This method must not be used outside a protocol adapter");
```

Comment:
IMHO, shadowing the (unsupported) parent static factory method with what is now in `createWithAnonymousLinkAddress` would be a better idea. The unused `replyId` parameter could be ignored and another `create` method could be added without the `replyId` parameter.

Reply:
That does not work because of incompatible return types.

Reply:
Ok, I see.